[Richard Li Defa] iP

rldefa opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ *SUPER FAST* to use

All you need to do is,
1. download it from here.
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

rldefa committed:
Greet, Echo and Exit

rldefa committed:
Add, List

rldefa committed:
Mark as Done

rldefa committed:
ToDos, Events, Deadlines

rldefa committed:
Automated Text UI Testing

rldefa committed:
Handle Errors

rldefa committed:
Delete

rldefa committed:
Write javadocs

rldefa committed:
Save

rldefa committed:
Dates and Times

rldefa committed:
Resolve merge conflict

rldefa committed:
Organize into Packages

rldefa committed:
Add JUnit Tests

rldefa committed:
Create a JAR File

rldefa committed:
JavaDoc

rldefa committed:
Follow the Coding Standard

rldefa committed:
Find

rldefa committed:
Merge branch 'branch-A-CodingStandard'
Merge changes to obey coding standards into master branch

rldefa committed:
Merge branch 'branch-Level-9'
Merge feature to find tasks into master branch

rldefa committed:
Merge branch 'master' into add-gradle-support
Merge all previous commits into adding gradle support

rldefa committed:
Use Gradle

rldefa committed:
Added CheckStyle

rldefa committed:
Moved main Duke class into duke package

phuccuongngo99 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,7 @@
+package duke;
+
+public class DukeException extends RuntimeException {
```
Can add in class header comments here :grin:

phuccuongngo99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,67 @@
+            newTask = new Todo(input.substring(5));
+        } else if (input.startsWith("deadline ")) {
+            int location = input.indexOf("/by ");
+            if (location < 0) throw new DukeException("Please provide a deadline.");
```
We should avoid one-liner if :thinking:

DystoriaX replied:
Yep agree. Even though it's one-liner, we still need to use the curly braces `{}`

phuccuongngo99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,67 @@
+     * @return the command object that represents the action that the user intends to make.
+     */
+    public static Command parse(String input) {
+        if (input.equals("bye")) return new ExitCommand();
```
Maybe switch-case would make the code easier to read and more compliant to coding standard :+1:

DystoriaX replied:
I'm not sure switch case can work here, as the author wants to use the method `.startsWith()` from the `String` object, but it is possible to do some small hacks. I can suggest two possible alternatives here:
- Preserve the if-else blocks, but comply with the coding standard
- You may look at `.split()` method to take the first command and use it into the switch statement

phuccuongngo99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,92 @@
+                    newTask = new Todo(taskString);
+                } else if (data.startsWith("D |")) {
+                    int dividerPosition = taskString.indexOf("|");
+                    newTask = new Deadline(taskString.substring(0, dividerPosition - 2), taskString.substring(dividerPosition + 2));
```
Different levels of abstraction in code here. 
Easier to read if we SLAP SLAP our code :grin: ([link](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#slap-hard))

phuccuongngo99 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+import java.util.stream.IntStream;
+
```
If you have time, can add class header comments and function comments :)

phuccuongngo99 reviewed (commented):
Very very clean code :+1: 
Minimal coding standard violation.
Keep up the nice work dude :grin:

DystoriaX commented on src/main/java/duke/DeleteCommand.java:
```diff
@@ -0,0 +1,44 @@
+     * @param storage custodian of reading and writing save files.
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        if (deleteIndex < 0 || deleteIndex >= tasks.size()) {
+            throw new DukeException("Task with that task number does not exist.");
+        }
+        Task deletedTask = tasks.delete(deleteIndex);
+        ui.announceTaskDeletion(deletedTask, tasks.size());
+        storage.save(tasks);
+    }
```
Perhaps you may want to add `throws DukeException` on the method. This helps to identify what kind of exceptions that are thrown by a method.

DystoriaX commented on src/main/java/duke/DoneCommand.java:
```diff
@@ -0,0 +1,43 @@
+     * @param storage custodian of reading and writing save files.
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        if (doneIndex < 0 || doneIndex >= tasks.size()) {
+            throw new DukeException("Task with that task number does not exist.");
+        }
+        ui.announceTaskCompletion(tasks.completeTask(doneIndex));
+        storage.save(tasks);
+    }
```
This too

DystoriaX commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,67 @@
+
+    private static DeleteCommand parseDelete(String input) {
+        int deletedTaskNumber;
+        try {
```
Maybe you want to add some space here? I usually separate variables and logic with one empty line. (Whichever suits your style, as my suggestion is subjective)

DystoriaX commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,67 @@
+            newTask = new Deadline(input.substring(9, location - 1), input.substring(location + 4));
+        } else if (input.startsWith("event ")) {
+            int location = input.indexOf("/at ");
+            if (location < 0) throw new DukeException("Please provide an event time.");
```
This too

DystoriaX reviewed (commented):
Only one coding standards that I can find (one-liner if). Other than that, LGTM. Really neat code.

rldefa committed:
Created GUI

rldefa committed:
Recover from backup

rldefa committed:
Use Assertions

rldefa committed:
Change an exclamation mark to full stop

rldefa committed:
Fix magic numbers

rldefa committed:
Merge pull request #3 from rldefa/branch-A-CodeQuality

Fix magic numbers

rldefa committed:
Resolve conflicts between Assertions and CodeQuality

rldefa committed:
Merge pull request #1 from rldefa/branch-A-Streams

Change an exclamation mark to full stop

rldefa committed:
Sort tasks based on description alphabetical order'

rldefa committed:
Merge branch 'master' of https://github.com/rldefa/ip

rldefa committed:
Storage: fix bug during load from save file

When events and deadlines were loaded form save file, the last character of the description string is missing.

This is due to the second argument of substring method being 1 too little.

rldefa committed:
Add screenshot of ui in docs

rldefa committed:
Set theme jekyll-theme-tactile

rldefa committed:
Find now returns filtered list instead of original list

rldefa committed:
Finish User Guide

rldefa committed:
Merge branch 'master' of https://github.com/rldefa/ip
Update local master with commits made by GitHub Pages

rldefa committed:
Resolve checkstyle warning

rldefa committed:
Fix Javadoc to use action verb in Storage and TaskList
